providers/siw/siw.c

```diff
@@ -0,0 +1,941 @@
+				sizeof(cmd));
+}
+
+static struct ibv_pd *siw_alloc_pd(struct ibv_context *ctx)
+{
+	struct ibv_alloc_pd cmd;
+	struct ib_uverbs_alloc_pd_resp resp;
+	struct ibv_pd *pd;
+
+	memset(&cmd, 0, sizeof(cmd));
+
+	pd = calloc(1, sizeof(*pd));
+	if (!pd)
+		return NULL;
+
+	if (ibv_cmd_alloc_pd(ctx, pd, &cmd, sizeof(cmd), &resp, sizeof(resp))) {
+		free(pd);
+		return NULL;
+	}
+	return pd;
+}
+
+static int siw_free_pd(struct ibv_pd *pd)
+{
+	int rv;
+
+	rv = ibv_cmd_dealloc_pd(pd);
+	if (rv)
+		return rv;
+
+	free(pd);
+	return 0;
+}
+
+static struct ibv_mr *siw_reg_mr(struct ibv_pd *pd, void *addr, size_t len,
+				 int access)
+{
+	struct siw_cmd_reg_umr_req cmd;
+	struct siw_cmd_reg_umr_resp resp;
+	struct siw_mr *mr;
+	int rv;
+
+	memset(&cmd, 0, sizeof(cmd));
+	memset(&resp, 0, sizeof(resp));
+
+	mr = calloc(1, sizeof(*mr));
+	if (!mr)
+		return NULL;
+
+	rv = ibv_cmd_reg_mr(pd, addr, len, (uintptr_t)addr, access,
+			    &mr->base_mr, &cmd.base, sizeof(cmd), &resp.base,
+			    sizeof(resp));
+	if (rv) {
+		free(mr);
+		return NULL;
+	}
+	return &mr->base_mr.ibv_mr;
+}
+
+static int siw_dereg_mr(struct verbs_mr *base_mr)
+{
+	struct siw_mr *mr = mr_base2siw(base_mr);
+	int rv;
+
+	rv = ibv_cmd_dereg_mr(base_mr);
+	if (rv)
+		return rv;
+
+	free(mr);
+	return 0;
+}
+
+static struct ibv_cq *siw_create_cq(struct ibv_context *ctx, int num_cqe,
+				    struct ibv_comp_channel *channel,
+				    int comp_vector)
+{
+	struct siw_cmd_create_cq cmd;
+	struct siw_cmd_create_cq_resp resp;
+	struct siw_cq *cq;
+	int rv;
+
+	memset(&cmd, 0, sizeof(cmd));
+	memset(&resp, 0, sizeof(resp));
+
+	cq = calloc(1, sizeof(*cq));
+	if (!cq)
+		return NULL;
+
+	rv = ibv_cmd_create_cq(ctx, num_cqe, channel, comp_vector, &cq->base_cq,
+			       &cmd.base, sizeof(cmd), &resp.base,
+			       sizeof(resp));
+	if (rv) {
+		if (siw_debug)
+			printf("libsiw: CQ creation failed: %d\n", rv);
+		free(cq);
+		return NULL;
+	}
+	pthread_spin_init(&cq->lock, PTHREAD_PROCESS_PRIVATE);
+
+	cq->id = resp.siw.cq_id;
+	cq->num_cqe = resp.siw.num_cqe;
+
+	if (resp.siw.cq_key != SIW_INVAL_UOBJ_KEY) {
+		int cq_size = resp.siw.num_cqe * sizeof(struct siw_cqe) +
+			      sizeof(struct siw_cq_ctrl);
+
+		cq->queue = mmap(NULL, cq_size, PROT_READ | PROT_WRITE,
+				 MAP_SHARED, ctx->cmd_fd, resp.siw.cq_key);
+
+		if (cq->queue == MAP_FAILED) {
+			if (siw_debug)
+				printf("libsiw: CQ mapping failed: %d", errno);
+			goto fail;
+		}
+		cq->ctrl = (struct siw_cq_ctrl *)&cq->queue[cq->num_cqe];
+		cq->ctrl->notify = SIW_NOTIFY_NOT;
+
+		return &cq->base_cq;
+	}
+	if (siw_debug)
+		printf("libsiw: prepare CQ mapping failed\n");
+fail:
+	ibv_cmd_destroy_cq(&cq->base_cq);
+	free(cq);
+
+	return NULL;
+}
+
+static int siw_resize_cq(struct ibv_cq *base_cq, int num_cqe)
+{
+	return -EOPNOTSUPP;
+}
+
+static int siw_destroy_cq(struct ibv_cq *base_cq)
+{
+	struct siw_cq *cq = cq_base2siw(base_cq);
+	int rv;
+
+	pthread_spin_lock(&cq->lock);
```

Comment:
Bit weird to hold the lock on something that is being freed? Does this do anything?

Reply:
OK will change that into pthread_spin_destroy(). wanted to make sure
nobody is still working on that CQ...

Reply:
Then use assert(pthread_try_lock) ?

Reply:
sounds good. i'll assert if not getting the lock, and destroy it after the ibv_cmd. If the ibv_cmd fails, I still have to unlock without destroy and return.